Source/Dafny/Compilers/Compiler-Csharp.cs

```diff
@@ -464,11 +471,80 @@ public CsharpCompiler(ErrorReporter reporter)
 
       CompileDatatypeDestructorsAndAddToInterface(dt, wr, interfaceTree, DtT_TypeArgs);
 
+      DtDowncastClone(dt, interfaceTree, nonGhostTypeArgs, @interface: true);
+      if (!dt.IsRecordType) {
+        DtDowncastClone(dt, wr, nonGhostTypeArgs);
+      }
+
       CompileDatatypeInterfaceMembers(dt, interfaceTree);
 
       return new ClassWriter(this, btw, null);
     }
 
+    private void DtDowncastClone(DatatypeDecl dt, ConcreteSyntaxTree wr, List<TypeParameter> nonGhostTypeArgs, bool @interface = false, bool lazy = false, DatatypeCtor ctor = null) {
+      if (nonGhostTypeArgs.Any(ty => ty.Variance == TypeParameter.TPVariance.Contra)) { return; };
+      var typeArgs = TypeParameters(nonGhostTypeArgs, safe: true);
+      var dtArgs = "_I" + dt.CompileName + typeArgs;
+      Func<(TypeParameter, int), string> PrintConverter = ((TypeParameter tArg, int i) t) => {
+        var name = IdName(t.tArg);
+        return $"Func<{name}, __{name}> converter{t.i}";
+      };
+      var argsWithIndex = nonGhostTypeArgs.Zip(Enumerable.Range(0, nonGhostTypeArgs.Count));
+
+      if (!@interface) {
+        wr.Write($"public {(ctor != null || lazy ? (dt.IsRecordType ? "" : "override ") : "abstract ")}");
+      }
+
+      wr.Write($"{dtArgs} DowncastClone{typeArgs}({argsWithIndex.Comma(PrintConverter)})");
+
+      if (ctor == null && !lazy) {
+        wr.WriteLine(";");
+        return;
+      }
+
+      var wBody = wr.NewBlock("");
+
+      wBody.WriteLine($"if (this is {dtArgs} dt) {{ return dt; }}");
+
+      Func<(string, Type), string> PrintInvocation = ((string name, Type ty) t) => {
+        var constructorIndex = -1;
+        Func<TypeParameter, Type, bool> ContainsTyVar = null;
+        ContainsTyVar = (tp, ty) => (ty.AsTypeParameter != null && ty.AsTypeParameter.Equals(tp))
+                                    || ty.TypeArgs.Exists(ty => ContainsTyVar(tp, ty));
+        if ((constructorIndex = nonGhostTypeArgs.IndexOf(t.ty.AsTypeParameter)) != -1) {
+          return $"converter{constructorIndex}({t.name})";
+        } else if (nonGhostTypeArgs.Exists(ty => ContainsTyVar(ty, t.ty))) {
+          var map = nonGhostTypeArgs.ToDictionary(tp => tp, tp =>
+            (Type)new UserDefinedType(tp.tok, new TypeParameter(tp.tok, $"_{tp.Name}", tp.VarianceSyntax)));
+          var sub = Resolver.SubstType(t.ty, map);
+          var downcast = new ConcreteSyntaxTree();
+          var @this = EmitDowncast(t.ty, sub, null, downcast);
+          @this.Write(t.name);
```

Comment:
Same here. It's confusing to see the keyword "this" even though it's a raw symbol.
Prefer the convention used in this file, such as `wr` or `wDownCastBody`.